flow/workflows/maintenance_flow.go

```diff
@@ -72,19 +76,59 @@ func StartMaintenanceWorkflow(ctx workflow.Context, input *protos.StartMaintenan
 }
 
 func startMaintenance(ctx workflow.Context, logger log.Logger) (*protos.StartMaintenanceFlowOutput, error) {
+	signalChan := model.StartMaintenanceSignal.GetSignalChannel(ctx)
+	maintenanceSelector := workflow.NewNamedSelector(ctx, "MaintenanceLoop")
+	skippedFlows := make(map[string]struct{})
+	cancelCurrentChild := func() {}
+
+	signalChan.AddToSelector(maintenanceSelector, func(maintenanceSignal *protos.StartMaintenanceSignal, _ bool) {
+		logger.Info("Received StartMaintenance Signal", slog.Any("signal", maintenanceSignal))
+		newSkippedFlows := make(map[string]struct{})
+		maps.Copy(newSkippedFlows, skippedFlows)
+		for _, flow := range maintenanceSignal.SkippedSnapshotWaitFlows {
+			newSkippedFlows[flow] = struct{}{}
+		}
+		skippedFlows = newSkippedFlows
+		cancelCurrentChild()
+	})
+	workflow.Go(ctx, func(ctx workflow.Context) {
+		for {
+			maintenanceSelector.Select(ctx)
+		}
+	})
+
 	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
 		StartToCloseTimeout: 24 * time.Hour,
 	})
 
 	snapshotWaitCtx := workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
-		StartToCloseTimeout: 24 * time.Hour,
+		StartToCloseTimeout: 2 * 24 * time.Hour,
 		HeartbeatTimeout:    1 * time.Minute,
+		RetryPolicy: &temporal.RetryPolicy{
+			MaximumInterval: 10 * time.Second,
+		},
 	})
-	waitSnapshotsFuture := workflow.ExecuteActivity(snapshotWaitCtx,
-		maintenance.WaitForRunningSnapshots,
-	)
-	err := waitSnapshotsFuture.Get(snapshotWaitCtx, nil)
-	if err != nil {
+	waitForSnapshotsWithSignalCheck := func() error {
+		for {
+			var snapshotWaitCancelCtx workflow.Context
+			snapshotWaitCancelCtx, cancelCurrentChild = workflow.WithCancel(snapshotWaitCtx)
+			waitSnapshotsFuture := workflow.ExecuteActivity(snapshotWaitCancelCtx,
+				maintenance.WaitForRunningSnapshots,
+				skippedFlows,
+			)
+			if err := waitSnapshotsFuture.Get(snapshotWaitCtx, nil); err != nil {
+				// It is ok to always ignore cancellation and always retry
+				// if more information is needed, we can maintain some state whether we cancelled it or someone else
+				if errors.Is(err, workflow.ErrCanceled) {
```

Comment:
should check if `snapshotWaitCtx` is canceled, in that case exit

Reply:
wouldn't cooperative cancellation always ensure that `snapshotWaitCtx` is also always cancelled if we take the `ErrCanceled` branch?